Added support for Xiaomi Philips LED Ceiling Lamp

kuduka opened the pull request:
Usage: miceil [OPTIONS] COMMAND [ARGS]...

  A tool to command Xiaomi Philips LED Ceiling Lamp.

Options:
  --ip TEXT
  --token TEXT
  -d, --debug
  --help        Show this message and exit.

Commands:
  acct_off    Auto CCT on.
  acct_on     Auto CCT on.
  delay_off   Set delay off in seconds.
  discover    Search for plugs in the network.
  off         Power off.
  on          Power on.
  set_bright  Set brightness level.
  set_cct     Set CCT level.
  set_scene   Set scene number.
  sml_off     Smart Midnight Light off.
  sml_on      Smart Midnight Light on.
  status      Returns the state information.

Not supported:

   - Auto On/Off Not Supported
   - Adjust Scens with Wall Switch Not Supported

rytilahti commented:
Hi @kuduka and thanks for the PR! It is looking code-wise good but I'm currently hesitant to add more console tools to be installed, so I think I'll have to make some time and make it easier to add new devices without having to copy & paste the cli tool code around.

Could you please do `mirobo -d info` with a recent device to gather the model information for this and the another eyecare lamp? I think the new tool should use the model information, and provide the device-specific commands to be defined in the device classes (e.g. Ceil in this case).

kuduka commented:
Sure, I'm no proficient with Python or even programming but I bet there's a better way to do it as we're replicating the same code over and over.

```
(venv) [kuduk@xps13 python-mirobo]$ venv/bin/mirobo --ip 192.168.1.71 --token 3d5f7ae53b51aa312e464b150b37453b -d info
INFO:mirobo.vacuum_cli:Debug mode active
ERROR:mirobo.vacuum_cli:Unable to read the stored msgid: [Errno 2] No such file or directory: '/tmp/python-mirobo.seq'
DEBUG:mirobo.vacuum_cli:Connecting to 192.168.1.71 with token 3d5f7ae53b51aa312e464b150b37453b
DEBUG:mirobo.protocol:Unable to decrypt, returning raw bytes.
DEBUG:mirobo.device:Discovered 884 55764 with ts: 1970-01-01 00:01:45
DEBUG:mirobo.device:192.168.1.71:54321 >>: {'id': 1, 'method': 'miIO.info', 'params': []}
DEBUG:mirobo.device:192.168.1.71:54321 (ts: 1970-01-01 00:01:45, id: 1) << {'result': {'life': 105, 'token': '3d5f7ae53b51aa312e464b150b37453b', 'mac': '34:CE:00:92:E6:21', 'fw_ver': '1.5.0', 'hw_ver': 'MTK7697', 'uid': '1647617138', 'model': 'philips.light.ceiling', 'mcu_fw_ver': '0027', 'wifi_fw_ver': '4.0.0', 'ap': {'rssi': -61, 'ssid': 'XarxaCasa', 'bssid': '5C:35:3B:EA:97:DE'}, 'netif': {'localIp': '192.168.1.71', 'mask': '255.255.255.0', 'gw': '192.168.1.1', 'gw_mac': '5C:35:3B:EA:97:DC'}, 'mmfree': 106488}, 'id': 1}
philips.light.ceiling v1.5.0 (34:CE:00:92:E6:21) @ 192.168.1.71 - token: 3d5f7ae53b51aa312e464b150b37453b
DEBUG:mirobo.vacuum_cli:Full response: {'ap': {'bssid': '5C:35:3B:EA:97:DE', 'rssi': -61, 'ssid': 'XarxaCasa'},
 'fw_ver': '1.5.0',
 'hw_ver': 'MTK7697',
 'life': 105,
 'mac': '34:CE:00:92:E6:21',
 'mcu_fw_ver': '0027',
 'mmfree': 106488,
 'model': 'philips.light.ceiling',
 'netif': {'gw': '192.168.1.1',
           'gw_mac': '5C:35:3B:EA:97:DC',
           'localIp': '192.168.1.71',
           'mask': '255.255.255.0'},
 'token': '3d5f7ae53b51aa312e464b150b37453b',
 'uid': '1647617138',
 'wifi_fw_ver': '4.0.0'}
DEBUG:mirobo.vacuum_cli:Writing {'seq': 1, 'manual_seq': 0} to /tmp/python-mirobo.seq
```

syssi commented:
The xiaomi philips led ball (9290012800, 450lm, 3000K-5700K, 6.5W, E27) supports the same feature set:

```
$ ./mirobo --ip 192.168.130.58 --token 6f3cxxxxxxxxxxxxxxxxxxxxxxxx3790 -d info 
INFO:mirobo.vacuum_cli:Debug mode active
DEBUG:mirobo.vacuum_cli:Read stored sequence ids: {'manual_seq': 0, 'seq': 1}
DEBUG:mirobo.vacuum_cli:Connecting to 192.168.130.58 with token 6f3xxxxxxxxxxxxxxxxxxxxxxxx13790
DEBUG:mirobo.protocol:Unable to decrypt, returning raw bytes.
DEBUG:mirobo.device:Discovered 926 2690 with ts: 1970-01-01 00:06:52
DEBUG:mirobo.device:192.168.130.58:54321 >>: {'id': 2, 'method': 'miIO.info', 'params': []}
DEBUG:mirobo.device:192.168.130.58:54321 (ts: 1970-01-01 00:06:52, id: 2) << {'id': 2, 'result': {'fw_ver': '1.3.0_0032', 'token': '6f3cxxxxxxxxxxxxxxxxxxxxxxxx3790', 'hw_ver': 'ESP8266', 'model': 'philips.light.bulb', 'life': 412, 'ap': {'rssi': -61, 'bssid': 'C0:25:06:A8:9A:5E', 'ssid': 'Greendale'}, 'mac': '34:CE:FF:FF:FF:FF', 'wifi_fw_ver': '1.5.0-dev(7f7a714)', 'mmfree': 5744, 'netif': {'gw': '192.168.130.1', 'mask': '255.255.255.0', 'localIp': '192.168.130.58'}}}
philips.light.bulb v1.3.0_0032 (34:CE:FF:FF:FF:FF) @ 192.168.130.58 - token: 6f3cxxxxxxxxxxxxxxxxxxxxxxxx3790
DEBUG:mirobo.vacuum_cli:Full response: {'ap': {'bssid': 'C0:25:06:A8:9A:5E', 'rssi': -61, 'ssid': 'Greendale'},
 'fw_ver': '1.3.0_0032',
 'hw_ver': 'ESP8266',
 'life': 412,
 'mac': '34:CE:FF:FF:FF:FF',
 'mmfree': 5744,
 'model': 'philips.light.bulb',
 'netif': {'gw': '192.168.130.1',
           'localIp': '192.168.130.58',
           'mask': '255.255.255.0'},
 'token': '6f3cxxxxxxxxxxxxxxxxxxxxxxxx3790',
 'wifi_fw_ver': '1.5.0-dev(7f7a714)'}
```

syssi commented:
I would be happy if this PR will be merged. Thanks in advance! :-)

rytilahti commented:
@syssi did you test it out? If everything is working I think we can just merge all the pending PRs as they are, and refactor the client tool(s) afterwards. The information necessary is in the PRs anyway, so if you say they're good to go.

rytilahti commented on mirobo/ceil.py:
```diff
@@ -0,0 +1,112 @@
+        """Auto CCT Off."""
+        return self.send("enable_ac", [0])
+
+    def status(self):
```
This could be set with typing to return `CeilStatus`, or is that deliberately left out to avoid moving it on top of this class?

syssi replied:
@kuduka Could to change this line to ` def status(self): -> CeilStatus` and move the class CeilStatus to the top of the file?

kuduka replied:
I've fixed but I think that I did a merge that wasn't needed on this PR, bare with me as I'm not a developer.

If I did something wrong, please let me know so I could improve it :)

syssi commented:
I was lazy and waiting for the merge. I will test it tonight.

syssi commented:
Just tested. Works fine! :+1:

kuduka commented:
cctsw is the property of "Adjust Scenes with Wall Switch" functionality that's not implemented yet on this code.

syssi commented:
Got it! Does your lamp provide a value for "cct" (color_temp)?

mirobo --ip addr --token token raw_command get_prop "['cct']"

It should be a value between 1 and 100. Which can be set by set_cct. Please provide this information if available for your device.

kuduka commented:
Yes, it does, so I've added it:

```
DEBUG:mirobo.device:192.168.1.227:54321 >>: {'id': 2, 'method': 'set_cct', 'params': [25]}
DEBUG:mirobo.device:192.168.1.227:54321 (ts: 1970-01-01 00:20:18, id: 2) << {'result': ['ok'], 'id': 2}

Sending cmd get_prop with params ['cct']
[25]

```
Also I've removed unused properties as I've just discovered that there's some kind of issue getting all properties at once, for example:

1) cctsw param if queried with sw it gets different results

```
# mirobo raw_command get_prop "'power', 'bl', 'cctsw'"
Sending cmd get_prop with params ('power', 'bl', 'cctsw')
['on', 1, [[0, 3], [0, 2], [0, 1]]]

# mirobo raw_command get_prop "'power', 'bl', 'sw'"
Sending cmd get_prop with params ('power', 'bl', 'sw')
['on', 1, [[0, 3], [0, 2], [0, 1]]]

# mirobo raw_command get_prop "'power', 'bl', 'cctsw', 'sw'"
Sending cmd get_prop with params ('power', 'bl', 'sw', 'cctsw')
['on', 1, [[0, 3], [0, 2], [0, 1]], 1]
```

2) If I query more than 8 params at once, last one gets overwrited or it disappears:

```
Sending cmd get_prop with params ('power', 'bl', 'ac', 'mb', 'ms', 'snm', 'bright', 'cct')
['on', 1, 0, 1, 1, 0, 100, 100]

Sending cmd get_prop with params ('power', 'bl', 'ac', 'mb', 'ms', 'snm', 'bright', 'cct', 'sw')
['on', 1, 0, 1, 1, 0, 100, [[0, 3], [0, 2], [0, 1]]]

Sending cmd get_prop with params ('power', 'bl', 'ac', 'mb', 'ms', 'sw', 'bright', 'cct')
['on', 1, 0, 1, 1, [[0, 3], [0, 2], [0, 1]], 100, 100]

Sending cmd get_prop with params ('power', 'bl', 'ac', 'mb', 'ms', 'sw', 'bright', 'cct', 'snm')
['on', 1, 0, 1, 1, [[0, 3], [0, 2], [0, 1]], 100, 100]
```

rytilahti commented:
Thanks a lot again @kuduka! Btw, in PRs it is better to use rebase than merge when rebasing after changes in the master:
```
git pull
<a merge commit gets created>
git rebase -i
<the log is without merge commits again>
```
This will allow keeping the history cleaner. It's not a big deal though, I'll merge this one too.

syssi commented:
Awesome! Please bump the version number.

rytilahti commented:
Could you do that & update the README.md to have a mention about other supported devices? I'm hesitating bumping the version before making the README up to date, but do not have currently time to make that happen (I'd like to fix the cli tool issue, but that may need to wait).

syssi commented:
I will prepare a PR!
